---
title: "Data Preparation"
author: "Keisuke Atsumi"
date: '`r format(Sys.time(), "%y/%m/%d %H:%M")`'
output:
  html_document:
    code_folding: hide
    theme: united
    toc: yes
    toc_float: yes
---

```{r setup, message = FALSE}

rm(list=ls())  # reset workspace
options(scipen=100)  # do not show numbers using exponential

# install & load packages
pacman::p_load(
  tidyverse,
  kableExtra,
  pander,     # nice tables
  openxlsx,   # read Excel file
  knitr,      # rmarkdown
  rotl,       # Open Tree of Life
  phytools,   # Phylogeny
  svglite,    # Output SVG
  devtools,   # github install
  magrittr,   # extend pipe
  ggimage,    # use pictures on ggtree
  patchwork   # combine multiple figure
)
opts_chunk$set(prompt=TRUE, message=FALSE, comment="", echo = FALSE, warning = FALSE) 

# githubinstall::githubinstall("ggtree")
library(ggtree)

```

# Data shaping

# Load data and pre-processing

```{r load unprocessed dat, results = 'asis'}

## Phenotype data ##
pheno <- read.xlsx(
    "../data/original.data.xlsx", sheet = "Phenotype"
    ) %>%
  # Excluding data without any parent data
  drop_na(contains("SD.sp")) %>%
  # to numeric
  mutate_at(vars(Mn.sp1:Bottleneck.sp2M), as.numeric) %>%
  # delete observation with negative mean trait value
  filter(Mn.sp1 > 0 & Mn.sp2 > 0 & homologous == "Yes") %>%   
  # calculate phenotypic difference between parental species
  mutate(
    pheno.div.ratio = abs(log(Mn.sp1/Mn.sp2))
    ) %>%
  mutate_at(vars(contains("pheno.div")), scale) %>%
  # describe whether full cross or not
  mutate(
    reciprocal = ifelse(
      is.na(Mn.hyb12 & Mn.hyb21), 
      "Inviable", "Viable"
      )
    )
#+++++++++++++++++++++++++#
# Metadata
#+++++++++++++++++++++++++#
meta <- read.xlsx(
    "../data/original.data.xlsx", sheet = "Metadata"
    ) %>%
  mutate_at(vars(contains("divergence")), as.numeric) %>%
  # Logalize genetic divergence
  mutate_at(vars(contains("divergence")), log10)
  

#+++++++++++++++++++++++++++++++++
# Combine phenotypic and metadata
#+++++++++++++++++++++++++++++++++
dat <- left_join(pheno, meta)
# add unique ID for each effect size (row)
dat$ES.ID <- paste("ES", sprintf("%03d",c(1:dim(dat)[1])), sep="")

#+++++++++++++++++++++++++++++++++
print("Scaling")
#+++++++++++++++++++++++++++++++++
print("Phenotypic divergence ratio")
attributes(pheno$pheno.div.ratio)
print("Minimum scaled value")
min(pheno$pheno.div.ratio)
print("Maximum scaled value")
max(pheno$pheno.div.ratio)

```

# summary figure
```{r summary figure, results = 'asis'}

#+++++++++++++++++++++#
# Phylogeny
#+++++++++++++++++++++#

# species pairs data 
taxa <- meta %>%
  select(-Cross.ID) %>%
  distinct_all(.keep_all = TRUE) %>%
  mutate_at("taxa", as.factor)

# matching names from open tree taxonomy
order <- tnrs_match_names(
  names = levels(taxa$taxa) %>%
    str_replace_all("_", " "), 
  context_name = "Animals"
  )

# which names return more than 1 match?
multimatch <- order$ott_id[order$number_matches != 1]
# inspect(order, ott_id = multimatch[1]) # Anura, confirmed adopted ott_id is correct
# inspect(order, ott_id = multimatch[2]) # Neuroptera, confirmed adopted ott_id is correct

# Create taxonomic tree
tree <- tol_induced_subtree(ott_ids = order$ott_id)
# Remove ott ids from tip label
tree$tip.label %<>%
  strip_ott_ids(remove_underscores=TRUE)
tree$tip.label[8] <- "Anura"

# # N of species pairs at each order
# factor(meta.sum$taxa, levels = as.vector(tree$tip.label))

plot.phylo <- ggtree(tree) + 
  geom_tiplab(
    geom = "image",
    image = paste("../data/pic", tree$tip.label, "png", sep = "."), 
    size = 0.1, offset = 3, hjust = 1
    ) +
  geom_tiplab(
    geom = "text",  
    label = tree$tip.label, 
    align = 1, size = 5
    ) +
  xlim(NA, 7)

# images from phylopic #    
# Lepidoptera (genus Catocala) "c224abfd-ee39-4923-98e5-c2606dcc56cb", 
# Diptera (Drosophila) "0cd6cc9f-683c-470e-a4a6-3b68beb826fa"
# Orthoptera (cricket) "b80d830b-155a-4ca5-9119-9a9fde019cc6"
# Rodentia (mouse), drawn by David Liao "0f6af3d8-49d2-4d75-8edf-08598387afde"
# Aves (Quail) "42f85a2b-7517-439a-8dc3-b745a35c035d"
# Anura (genus Hyla) "A22D353D-B045-4044-9432-8F340B9A3104", 
# Cichliformes (Nile tilapia), drawn by Milton Tan "84c7e672-2593-44a6-a807-cffbd3156cc5" 


#+++++++++++++++++++++#
# Summary 
#+++++++++++++++++++++#

summary <- dat %>%
  group_by(taxa) %>%
  summarise(
    'Species pair' = length(unique(species.pair)),
    'Obser- vation' = length(unique(ES.ID)),
    Study = length(unique(Study.ID))
    ) %>%
  # Order taxon so that match with phylogenetic tree topology
  within(taxa <- ordered(taxa, levels = c("Neuroptera", "Coleoptera", "Diptera", "Lepidoptera", "Orthoptera", "Aves", "Rodentia", "Anura", "Cichliformes")))


### Table ###
summary %>%
  summarise_if(is.numeric, sum) %>%
    mutate(taxa = "Total") %>%
  bind_rows(summary, .) %>%
  kable("html", digits = 3) %>% 
  kable_styling("striped", position = "left")

### Plot ###
plot.sum <- ggplot(
  # Make tidy summary data
  summary %>%
  gather(key = metrics, value = N, -c(taxa)) %>%
  # Order metrics
  within(
    metrics <- ordered(
      metrics,
      levels = c("Study", "Species pair", "Obser- vation")
      )
    ), 
  aes(x = "", y = N, fill = taxa)) +
  geom_bar(width = 1, stat = "identity") +
  scale_fill_brewer(palette = "YlGn") +
  facet_wrap(
    scale = "free", "metrics",
    # strip text into two lines
    labeller = label_wrap_gen(width = 8)
    ) +
  xlab("") +
  theme(
    strip.text = element_text(size = 15),
    axis.text.y = element_text(size = 14),
    panel.grid.minor = element_blank(),
    panel.grid.major.x = element_blank(), 
    axis.ticks.x = element_blank(),
    legend.text = element_blank(),
    legend.title = element_blank()
    )

#+++++++++++++++++#
# Combine figures
#+++++++++++++++++#
plot.all <- plot.phylo + plot.sum
plot.all
# ggsave(
#   plot = plot.all, 
#   file = "../Analysis/data.svg", height = 5, width = 8.5
#   )

```

# Metadata

```{r, results='asis'}

metavars <- c("reciprocal", "distribution", "hetero.sex")

meta.unique <- dat %>%
  distinct_at("species.pair", .keep_all = TRUE) %>%
  mutate_at(metavars, as.factor)

meta.all.sp <- meta.unique %>% 
  group_by(taxa) %>%
  summarise(Total = length(species.pair)) 

for (i in metavars) {
  
  bind_rows(
    # At each taxon
    meta.unique %>% 
      group_by_("taxa", i) %>%
      summarise(species = length(species.pair))
    ,
    # All taxon
    meta.unique %>% 
      group_by_(i) %>%
      summarise(species = length(unique(species.pair))) %>% 
      mutate(taxa = "Total")
    ) %>%
    ungroup %>%
    spread(key = i, value = "species") %>%
    mutate_all(~replace(., is.na(.), 0)) %>% # replace NA by 0
    as.data.frame %>%
    # All species pair at each taxon: summing columns
    mutate(Total = rowSums(.[2:(1+nlevels(meta.unique[, i]))])) %>%
    # Calc % of level 2 (i.e. male hetero, ...)
    mutate('%' = round(.[3]/Total*100, 1)) %>%
    rename(!!paste0(levels(meta.unique[, i])[2], "%") := "%") %>%
    # Table output
    kable("html", digits = 3, caption = i) %>% 
    kable_styling("striped", position = "left") %>% 
    print()

}  
  

```


# Data used in meta-regression (all metadata available)

```{r}

full_join(
  # All species pair
  dat %>%
    select(c("taxa", "species.pair")) %>%
    distinct_all() %>%
    group_by(taxa) %>%
    summarise('All species' = length(species.pair)),
  # Species pair used in meta-regression
  dat %>%
    filter(!is.na(sex.determination), !is.na(divergence.COI)) %>%
    select(c("taxa", "species.pair")) %>%
    distinct_all() %>%
    group_by(taxa) %>%
    summarise('Metadata available' = length(species.pair))
  ) %>%
  mutate_all(~replace(., is.na(.), 0)) %>% # replace NA by 0
  kable("html") %>% 
  kable_styling("striped", position = "left")

```



# Allign parental species according to trait values
Currently, dataset is composed of   
species 1, species 2, hybrid12 (sp1 mother & sp2 father), hybrid 21 (sp2 mother & sp1 father)  

To investigate paternal effect in hybrids using difference in mean trait value between reciprocal cross, it's better to allign species with larger trait value (spL) and smaller value (spS) in each traits. Name of hybrids will be changed according to change in names of parental species (hybridLS & hybridSL).  
If   
ln(hybridLS/hybridSL) or (hybridLS - hybridSL) >0, that means hybridLS > hybrid SL ...maternal effect  
ln(hybridLS/hybridSL) or (hybridLS - hybridSL) <0, that means hybridLS < hybrid SL ...paternal effect

```{r dat preprocessing for mean}

### define parental species by phenotypic mean : sp1 & sp2 -> spL & spS ###

############ in the case where sp1 >= sp2 ############
#### extract
sp1sp2.mean <- dat %>%
  filter(Mn.sp1 >= Mn.sp2) # filtering according to mean trait value of parentals
#### change column name from sp1 or sp2 -> spL or spS
names(sp1sp2.mean) <- gsub("1", "L", names(sp1sp2.mean))
names(sp1sp2.mean) <- gsub("2", "S", names(sp1sp2.mean))

############ in the case where sp1 < sp2 ############
#### extract
sp2sp1.mean <- dat %>%
  filter(Mn.sp1 < Mn.sp2)
#### change column name from sp1 or sp2 -> spS or spL
names(sp2sp1.mean) <- gsub("1", "S", names(sp2sp1.mean))
names(sp2sp1.mean) <- gsub("2", "L", names(sp2sp1.mean))

############ combine dataset & sort by ES.ID ############
bind_rows(sp1sp2.mean, sp2sp1.mean) %>%
  arrange(ES.ID) %>%
  write.csv("../data/dat.mean.csv", quote=F,row.names = F)

```

# Allign parental species according to phenotypic variation (CV)
Currently, dataset is composed of   
species 1, species 2, hybrid12 (sp1 mother & sp2 father), hybrid 21 (sp2 mother & sp1 father)  

To investigate paternal effect in hybrids using difference in coefficient of variation (CV) between reciprocal cross, it's better to allign species with larger CV (spL) and smaller CV (spS) in each traits. Name of hybrids will be changed according to change in names of parental species (hybridLS & hybridSL).  
If   
ln(hybridLS/hybridSL) or (hybridLS - hybridSL) >0, that means hybridLS > hybrid SL ...maternal effect  
ln(hybridLS/hybridSL) or (hybridLS - hybridSL) <0, that means hybridLS < hybrid SL ...paternal effect

```{r dat preprocessing for variation}

### define parental species by CV : sp1 & sp2 -> spL & spS ###

############ in the case where sp1 >= sp2 ############
#### extract
sp1sp2.var <- dat %>%
  filter(SD.sp1/Mn.sp1 >= SD.sp2/Mn.sp2) # filtering according to CV of parentals
#### change column name from sp1 or sp2 -> spL or spS
names(sp1sp2.var) <- gsub("1", "L", names(sp1sp2.var))
names(sp1sp2.var) <- gsub("2", "S", names(sp1sp2.var))

############ in the case where sp1 < sp2 ############
#### extract
sp2sp1.var <- dat %>%
  filter(SD.sp1/Mn.sp1 < SD.sp2/Mn.sp2)
#### change column name from sp1 or sp2 -> spS or spL
names(sp2sp1.var) <- gsub("1", "S", names(sp2sp1.var))
names(sp2sp1.var) <- gsub("2", "L", names(sp2sp1.var))

############ combine dataset & sort by ES.ID ############
bind_rows(sp1sp2.var, sp2sp1.var) %>%
  arrange(ES.ID) %>%
  write.csv("../data/dat.CV.csv", quote=F,row.names = F)

```

# Another dataset to investigate relationship between novel phenotype and novel variability

```{r}

transgression <- bind_rows(
  # hyb12
  dat %>%
    drop_na(Mn.hyb12, SD.hyb12) %>%
    mutate(
      cross = "1x2",
      mean.TS = ifelse(
        # Exceeding upper range: 1
        Mn.hyb12 > Mn.sp1 & Mn.hyb12 > Mn.sp2,
        1,
        ifelse(
          # Exceeding lower range: 1
          Mn.hyb12 < Mn.sp1 & Mn.hyb12 < Mn.sp2,
          -1,
          0
        )
      ),
      var.TS = ifelse(
        # Exceeding upper range: 1
        SD.hyb12/Mn.hyb12 > SD.sp1/Mn.sp1 & 
          SD.hyb12/Mn.hyb12 > SD.sp2/Mn.sp2,
        1,
        ifelse(
          # Exceeding lower range: 1
          SD.hyb12/Mn.hyb12 < SD.sp1/Mn.sp1 & 
            SD.hyb12/Mn.hyb12 < SD.sp2/Mn.sp2,
          -1,
          0
        )
      )    
    ),
  # hyb21
  dat %>%
    drop_na(Mn.hyb21, SD.hyb21) %>%
    mutate(
      cross = "2x1",
      mean.TS = ifelse(
        # Exceeding upper range: 1
        Mn.hyb21 > Mn.sp1 & Mn.hyb21 > Mn.sp2,
        1,
        ifelse(
          # Exceeding lower range: 1
          Mn.hyb21 < Mn.sp1 & Mn.hyb21 < Mn.sp2,
          -1,
          0
        )
      ),
      var.TS = ifelse(
        # Exceeding upper range: 1
        SD.hyb21/Mn.hyb21 > SD.sp1/Mn.sp1 & 
          SD.hyb21/Mn.hyb21 > SD.sp2/Mn.sp2,
        1,
        ifelse(
          # Exceeding lower range: 1
          SD.hyb21/Mn.hyb21 < SD.sp1/Mn.sp1 & 
            SD.hyb21/Mn.hyb21 < SD.sp2/Mn.sp2,
          -1,
          0
        )
      )    
    ) 
  ) %>%
  mutate(mean.TS.occur = ifelse(mean.TS == 0, 0, 1)) %>%
  write.csv("../data/dat.transgression.csv", quote=F,row.names = F)


```

# Session information
```{r}
sessionInfo()  # show R version etc
```

